File: src/main/java/com/fish/extendedae_plus_client/integration/recipeViewer/jei/ViewerJei.java
package com.fish.extendedae_plus_client.integration.recipeViewer.jei;

import appeng.api.stacks.AEFluidKey;
import appeng.api.stacks.AEItemKey;
import appeng.api.stacks.AEKey;
import appeng.api.stacks.GenericStack;
import com.fish.extendedae_plus_client.integration.ContextModLoaded;
import com.fish.extendedae_plus_client.integration.recipeViewer.IRecipeViewer;
import com.mojang.datafixers.util.Pair;
import mezz.jei.api.constants.VanillaTypes;
import mezz.jei.api.ingredients.IIngredientType;
import mezz.jei.api.ingredients.ITypedIngredient;
import mezz.jei.api.neoforge.NeoForgeTypes;
import net.minecraft.client.gui.screens.Screen;
import org.jetbrains.annotations.Nullable;
import org.lwjgl.glfw.GLFW;
import tamaized.ae2jeiintegration.integration.modules.jei.GenericEntryStackHelper;

import java.util.Collections;
import java.util.List;

public class ViewerJei implements IRecipeViewer {
    @Override
    public List<GenericStack> getHoveredStacks(double mouseX, double mouseY) {
        return getHoveredStacks();
    }

    @Override
    public List<GenericStack> getHoveredStacks() {
        ITypedIngredient<?> hovered = HelperJeiRuntime.getIngredientUnderMouse().orElse(null);
        if (hovered != null)
            return Collections.singletonList(GenericEntryStackHelper.ingredientToStack(hovered));
        else return null;
    }

    @Override
    public List<GenericStack> getFavorites() {
        return HelperJeiRuntime.getBookmarkList().stream()
                .map(GenericEntryStackHelper::ingredientToStack).toList();
    }

    @Override
    public @Nullable Pair<Boolean, Boolean> matchesKey(int mouseKey) {
        boolean match = false;
        boolean stack = false;
        boolean toInv = true;

        if (Screen.hasControlDown()) match = true;
        if (match && Screen.hasShiftDown()) stack = true;

        if (mouseKey == GLFW.GLFW_MOUSE_BUTTON_RIGHT) toInv = false;

        return match
                ? new Pair<>(stack, toInv)
                : null;
    }

    @Override
    public boolean isCheatMode() {
        return HelperJeiRuntime.isJeiCheatModeEnabled();
    }

    @Override
    public void addFavorite(GenericStack stack) {
        AEKey key = stack.what();
        if (key instanceof AEItemKey itemKey)
            HelperJeiRuntime.addFavorite(itemKey.toStack(), VanillaTypes.ITEM_STACK);
        else if (key instanceof AEFluidKey fluidKey)
            HelperJeiRuntime.addFavorite(fluidKey.toStack(1000), NeoForgeTypes.FLUID_STACK);
        else if (ContextModLoaded.mekanism.isLoaded() && ContextModLoaded.appliedMekanistics.isLoaded()) {
            try {
                var clazzTypeKey = Class.forName("me.ramidzkh.mekae2.ae2.MekanismKey");
                if (!clazzTypeKey.isAssignableFrom(key.getClass())) return;

                var fieldStack = clazzTypeKey.getMethod("getStack").invoke(key);

                @SuppressWarnings("unchecked")
                var fieldTypeIngredient = (IIngredientType<Object>)
                        Class.forName("mekanism.client.recipe_viewer.jei.MekanismJEI")
                        .getField("TYPE_CHEMICAL").get(null);

                HelperJeiRuntime.addFavorite(fieldStack, fieldTypeIngredient);
            } catch (Exception ignored) {}
        }
    }

    @Override
    public void setSearch(String text) {
        HelperJeiRuntime.setIngredientFilterText(text);
    }
}

File: src/main/kotlin/com/fish/extendedae_plus_client/integration/recipeViewer/jei/ViewerJei.kt
package com.fish.extendedae_plus_client.integration.recipeViewer.jei

import appeng.api.stacks.AEFluidKey
import appeng.api.stacks.AEItemKey
import appeng.api.stacks.GenericStack
import appeng.integration.modules.jei.GenericEntryStackHelper
import com.fish.extendedae_plus_client.integration.ContextModLoaded
import com.fish.extendedae_plus_client.integration.recipeViewer.IRecipeViewer
import com.mojang.datafixers.util.Pair
import mezz.jei.api.constants.VanillaTypes
import mezz.jei.api.forge.ForgeTypes
import mezz.jei.api.ingredients.IIngredientType
import net.minecraft.client.gui.screens.Screen
import org.lwjgl.glfw.GLFW

class ViewerJei : IRecipeViewer {
    override fun getHoveredStacks(mouseX: Double, mouseY: Double): MutableList<GenericStack?> {
        return hoveredStacks
    }

    override val hoveredStacks: MutableList<GenericStack?>
        get() {
            val hovered = HelperJeiRuntime.ingredientUnderMouse.orElse(null)
            return if (hovered != null) mutableListOf(GenericEntryStackHelper.ingredientToStack(hovered))
            else mutableListOf()
        }

    override val favorites: MutableList<GenericStack?>
        get() = HelperJeiRuntime.bookmarkList.stream()
            .map { ingredient -> GenericEntryStackHelper.ingredientToStack(ingredient) }
            .toList()

    override fun matchesKey(mouseKey: Int): Pair<Boolean, Boolean>? {
        var match = false
        var stack = false
        var toInv = true

        if (Screen.hasControlDown()) match = true
        if (match && Screen.hasShiftDown()) stack = true

        if (mouseKey == GLFW.GLFW_MOUSE_BUTTON_RIGHT) toInv = false

        return if (match)
            Pair<Boolean, Boolean>(stack, toInv)
        else
            null
    }

    override val isCheatMode: Boolean
        get() = HelperJeiRuntime.jeiCheatModeEnabled

    override fun addFavorite(stack: GenericStack) {
        val key = stack.what()
        if (key is AEItemKey) HelperJeiRuntime.addFavorite(key.toStack(), VanillaTypes.ITEM_STACK)
        else if (key is AEFluidKey) HelperJeiRuntime.addFavorite(
            key.toStack(1000),
            ForgeTypes.FLUID_STACK
        )
        else if (ContextModLoaded.mekanism.isLoaded && ContextModLoaded.appliedMekanistics.isLoaded) {
            try {
                val clazzTypeKey = Class.forName("me.ramidzkh.mekae2.ae2.MekanismKey")
                if (!clazzTypeKey.isAssignableFrom(key.javaClass)) return

                val fieldStack = clazzTypeKey.getMethod("getStack").invoke(key)

                val fieldTypeIngredient = Class.forName("mekanism.client.recipe_viewer.jei.MekanismJEI")
                    .getField("TYPE_CHEMICAL").get(null) as IIngredientType<Any>

                HelperJeiRuntime.addFavorite(fieldStack, fieldTypeIngredient)
            } catch (_: Exception) {
            }
        }
    }

    override fun setSearch(text: String) {
        HelperJeiRuntime.setIngredientFilterText(text)
    }
}
